source/_posts/vue-design-patterns.md

```diff
@@ -0,0 +1,982 @@
+          .addField({
+            component: VInput,
+            name: "password",
+            type: "password",
+            label: "password",
+            props: {
+              value: "default password",
+            },
+            validation: z.string().min(10),
+          })
+          .build();
+      }
+    }
+```
+
+To validate our forms, we are using the [zod](https://zod.dev/) library. As we can see, `makeLoginForm()` uses the Form Builder class to create a new app form with a detailed description of each field.
+
+> Code:  [vue design pattern repository](https://github.com/cr0wg4n/vue-design-patterns), execute, and go to **/builder-pattern** route 😎.
+
+
+## Adapter Pattern
+The adapter pattern is a structural design pattern also known as "Wrapper". This pattern allows the interfacing of an existing class used as another interface, like something in the middle of a connection. The perfect real-world example is when you go to another country, you realize that the power plug is different. So, you will need an adapter to get power. That's the magic.
+
+A typical use case is to wrap a 3rd party library, which is used everywhere in your code. Someday, many libraries will be deprecated, or in the worst case, security issues, also known as vulnerabilities. To learn more about it, we have an amazing academy, [https://websecurity-academy.com/](https://websecurity-academy.com/), and a free guide related to [Vue Security Best Practices](https://dev-academy.com/vue-security-best-practices/).
+
+In this case, we are using the [js-cookie](https://github.com/js-cookie/js-cookie) library to interact directly with our Cookies, which is simple and lightweight. The adapter looks like this.
+
+```ts
+    // File: CookiesAdapter.ts
+    
+    import Cookies from "js-cookie";
+    
+    export interface CookieOptions {
+      expires?: number | Date | undefined;
+      path?: string | undefined;
+      domain?: string | undefined;
+      secure?: boolean | undefined;
+      sameSite?: "strict" | "lax" | "none" | undefined;
+      [property: string]: any;
+    }
+    
+    export default class CookiesAdapter {
+      cookies;
+    
+      constructor() {
+        this.cookies = Cookies;
+      }
+    
+      get(key: string): string | undefined {
+        return this.cookies.get(key);
+      }
+    
+      getAll(): object {
+        return this.cookies.get();
+      }
+    
+      set(
+        key: string,
+        value: string,
+        options: CookieOptions | undefined = undefined
+      ): string | undefined {
+        return this.cookies.set(key, value, options);
+      }
+    
+      remove(key: string) {
+        this.cookies.remove(key);
+      }
+    }    
+```
+
+If one day we will need to replace the 3rd party library or add more features, it will be fast, and many parts of your code won't be affected by this change, as we can see in the following schematic image.
+
+{% img "00-adapter-pattern.jpg" "Adapter pattern schema" "lazy" %}
```

Comment:
The image is not mobile friendly (readability) but let it be.
@bartosz-io We'll need to think about eg. responsive image tag in hexojs. It'd be useful to have it for small devices >= 360 & =< 420